src/circuit/client.ts

```diff
@@ -0,0 +1,362 @@
+import { logger } from '@libp2p/logger'
+import { RELAY_V2_HOP_CODEC } from './multicodec.js'
+import { getExpiration, namespaceToCid } from './utils.js'
+import {
+  CIRCUIT_PROTO_CODE,
+  DEFAULT_MAX_RESERVATIONS,
+  RELAY_RENDEZVOUS_NS
+} from './constants.js'
+import type { PeerId } from '@libp2p/interface-peer-id'
+import type { AddressSorter } from '@libp2p/interface-peer-store'
+import type { Connection } from '@libp2p/interface-connection'
+import sort from 'it-sort'
+import all from 'it-all'
+import { pipe } from 'it-pipe'
+import { publicAddressesFirst } from '@libp2p/utils/address-sort'
+import { reserve } from './v2/index.js'
+import { EventEmitter, CustomEvent } from '@libp2p/interfaces/events'
+import type { Startable } from '@libp2p/interfaces/startable'
+import type { Components } from '../components.js'
+import type { RelayReservationManagerConfig } from './index.js'
+
+const log = logger('libp2p:circuit:client')
+
+const noop = () => { }
+
+/**
+ * CircuitServiceInit initializes the circuit service using values
+ * from the provided config and an @type{AddressSorter}.
+ */
+export interface CircuitServiceInit extends RelayReservationManagerConfig {
+  /**
+   * Allows prioritizing addresses from the peerstore for dialing. The
+   * default behavior is to prioritise public addresses.
+   */
+  addressSorter?: AddressSorter
+  /**
+   * A callback to invoke when an error occurs in the circuit service.
+   */
+  onError?: (error: Error, msg?: string) => void
+}
+
+export interface RelayReservationManagerEvents {
+  'relay:reservation': CustomEvent
+}
+
+/**
+ * ReservationManager automatically makes a circuit v2 reservation on any connected
+ * peers that support the circuit v2 HOP protocol.
+ */
+export class RelayReservationManager extends EventEmitter<RelayReservationManagerEvents> implements Startable {
+  private readonly components: Components
+  private readonly addressSorter: AddressSorter
+  private readonly maxReservations: number
+  private readonly relays: Set<string>
+  private readonly reservationMap: Map<PeerId, ReturnType<typeof setTimeout>>
```

Comment:
Using PeerId objects as a key can be unsafe as they don't pass SameValueZero equality.

E.g.:

```ts
const peerString = 'a-peer-id-string'

const p1 = peerIdFromString(peerString)
const p2 = peerIdFromString(peerString)

console.info(p1 === p2) // false
```

Use `PeerMap` from [@libp2p/peer-collections](https://www.npmjs.com/package/@libp2p/peer-collections) instead.